File: lib/lib_timesys/j2date.f
     
C Procedure           J2DATE
C
      SUBROUTINE      J2DATE( JD, YEAR, MONTH, DAY )
C
C
C                            Purpose
C
C  This subroutine (Julian date 2 calendar DATE) converts the input inte
C  Julian date to the corresponding Gregorian calendar date. Since the J
C  date is an integer, this correspondence is exact for noon of the cale
C  date.
C
C  The algorithm for this conversion is taken from the following article
C  Tantzen,R.T., "Communications of the ACM", Volume 6, Number 8, August
C  Algorithm 199, page 444.
C
C
C                            Input_Arguments
C
C  JD       is the integer Julian date.
C
C
C                            Output_Arguments
C
C  YEAR     is the year number.
C  MONTH    is the month number.
C  DAY      is the day number.
C
C
C                            Output_Arguments
C
C                            Declarations_of_External_Functions
C
C                            Declarations_of_Input_and_Output_Arguments
C
      INTEGER            JD
      INTEGER            YEAR
      INTEGER            MONTH
      INTEGER            DAY
C
C
C                            Declarations_of_Local_Variables
C
      INTEGER            J
      INTEGER            Y
      INTEGER            M
      INTEGER            D
C
C
C                     Method
C-&
C***********************************************************************
C
      J = JD
C
C
      J = J - 1721119
      Y = (4*J-1)/146097
      J = 4*J - 1 - 146097*Y
      D = J/4
      J = (4*D+3)/1461
      D = 4*D + 3 -1461*J
      D = (D+4)/4
      M = (5*D-3)/153
      D = 5*D - 3 - 153*M
      D = (D+5)/5
      Y = 100*Y + J
      IF( M .LT. 10 )THEN
         M = M + 3
      ELSE
         M = M - 9
         Y = Y + 1
      END IF
C
C
      YEAR   = Y
      MONTH  = M
      DAY    = D
C
      RETURN
      END 

File: tool/GraceRead_L1_2008_03_20/lib/j2date.c
/* 

Purpose: 
 
   This subroutine (Julian date 2 calendar DATE) converts the input inte
   Julian date to the corresponding Gregorian calendar date. Since the J
   date is an integer, this correspondence is exact for noon of the cale
   date.
 
   The algorithm for this conversion is taken from the following article
   Tantzen,R.T., "Communications of the ACM", Volume 6, Number 8, August
   Algorithm 199, page 444.  

   Adapted from timetrans J2DATE.F
   Bruce Haines 3/5/96

*/

#include "TimeLib.h"

void j2date(

/* Input: */
		int jd,      /* is the integer julian data */
/* Output */

                int *year,
		int *month,
		int *day
	   )
{

/* Local: */
  static char SccsId[] = "@(#) j2date.c       1.1 01/08/01";

  int j;
  int y;
  int m;
  int d;

  j = jd;
  j = j - 1721119;
  y = (4*j-1)/146097;
  j = 4*j - 1 - 146097*y;
  d = j/4;
  j = (4*d+3)/1461;
  d = 4*d + 3 -1461*j;
  d = (d+4)/4;
  m = (5*d-3)/153;
  d = 5*d - 3 - 153*m;
  d = (d+5)/5;
  y = 100*y + j;
  if (m < 10) {
     m = m + 3;
  } else {
     m = m - 9;
     y = y + 1;
  }
  *year   = y;
  *month  = m;
  *day    = d;

}
